# Libarary


In [116]:
import pandas as pd
import seaborn as sns

import matplotlib.pyplot as plt
import numpy as np
import datetime as dt

# Data loading and cleaning

### 1. Transaction DF:

In [117]:
transaction_df = pd.read_excel('KPMG_VI_New_raw_data_update_final.xlsx','Transactions')

In [118]:
transaction_df

Note: The data and information in this document is reflective of a hypothetical situation and client. This document is to be used for KPMG Virtual Internship purposes only.   \
0                                         transaction_id                                                                                                                              
1                                                      1                                                                                                                              
2                                                      2                                                                                                                              
3                                                      3                                                                                                                              
4                                                      4                                                                                                                              
...                                                  ...                                                                                                                              
19996                                              19996                                                                                                                              
19997                                              19997                                                                                                                              
19998                                              19998                                                                                                                              
19999                                              19999                                                                                                                              
20000                                              20000                                                                                                                              

       Unnamed: 1   Unnamed: 2           Unnamed: 3    Unnamed: 4  \
0      product_id  customer_id     transaction_date  online_order   
1               2         2950  2017-02-25 00:00:00         False   
2               3         3120  2017-05-21 00:00:00          True   
3              37          402  2017-10-16 00:00:00         False   
4              88         3135  2017-08-31 00:00:00         False   
...           ...          ...                  ...           ...   
19996          51         1018  2017-06-24 00:00:00          True   
19997          41          127  2017-11-09 00:00:00          True   
19998          87         2284  2017-04-14 00:00:00          True   
19999           6         2764  2017-07-03 00:00:00         False   
20000          11         1144  2017-09-22 00:00:00          True   

         Unnamed: 5      Unnamed: 6    Unnamed: 7     Unnamed: 8  \
0      order_status           brand  product_line  product_class   
1          Approved           Solex      Standard         medium   
2          Approved   Trek Bicycles      Standard         medium   
3          Approved      OHM Cycles      Standard            low   
4          Approved  Norco Bicycles      Standard         medium   
...             ...             ...           ...            ...   
19996      Approved      OHM Cycles      Standard           high   
19997      Approved           Solex          Road         medium   
19998      Approved      OHM Cycles      Standard         medium   
19999      Approved      OHM Cycles      Standard           high   
20000      Approved   Trek Bicycles      Standard         medium   

         Unnamed: 9 Unnamed: 10    Unnamed: 11              Unnamed: 12  
0      product_size  list_price  standard_cost  product_first_sold_date  
1            me

In [119]:
# since there are no correct column names, use .iloc to use the current first row to replace the incorrect column names.

transaction_df.columns = transaction_df.iloc[0]

transaction_df.head()



0  transaction_id  product_id  customer_id     transaction_date  online_order  \
0  transaction_id  product_id  customer_id     transaction_date  online_order   
1               1           2         2950  2017-02-25 00:00:00         False   
2               2           3         3120  2017-05-21 00:00:00          True   
3               3          37          402  2017-10-16 00:00:00         False   
4               4          88         3135  2017-08-31 00:00:00         False   

0  order_status           brand  product_line  product_class  product_size  \
0  order_status           brand  product_line  product_class  product_size   
1      Approved           Solex      Standard         medium        medium   
2      Approved   Trek Bicycles      Standard         medium         large   
3      Approved      OHM Cycles      Standard            low        medium   
4      Approved  Norco Bicycles      Standard         medium        medium   

0  list_price  standard_cost  product_first_sold_date  
0  list_price  standard_cost  product_first_sold_date  
1       71.49          53.62                    41245  
2     2091.47         388.92                    41701  
3     1793.43         248.82                    36361  
4     1198.46          381.1                    36145

In [120]:
# Then, drop the duplicated first row (which is index 0)

transaction_df = transaction_df.drop(0)
transaction_df.head()

0 transaction_id product_id customer_id     transaction_date online_order  \
1              1          2        2950  2017-02-25 00:00:00        False   
2              2          3        3120  2017-05-21 00:00:00         True   
3              3         37         402  2017-10-16 00:00:00        False   
4              4         88        3135  2017-08-31 00:00:00        False   
5              5         78         787  2017-10-01 00:00:00         True   

0 order_status           brand product_line product_class product_size  \
1     Approved           Solex     Standard        medium       medium   
2     Approved   Trek Bicycles     Standard        medium        large   
3     Approved      OHM Cycles     Standard           low       medium   
4     Approved  Norco Bicycles     Standard        medium       medium   
5     Approved  Giant Bicycles     Standard        medium        large   

0 list_price standard_cost product_first_sold_date  
1      71.49         53.62                   41245  
2    2091.47        388.92                   41701  
3    1793.43        248.82                   36361  
4    1198.46         381.1                   36145  
5     1765.3        709.48                   42226

In [121]:
transaction_df.reset_index() 
transaction_df

0     transaction_id product_id customer_id     transaction_date online_order  \
1                  1          2        2950  2017-02-25 00:00:00        False   
2                  2          3        3120  2017-05-21 00:00:00         True   
3                  3         37         402  2017-10-16 00:00:00        False   
4                  4         88        3135  2017-08-31 00:00:00        False   
5                  5         78         787  2017-10-01 00:00:00         True   
...              ...        ...         ...                  ...          ...   
19996          19996         51        1018  2017-06-24 00:00:00         True   
19997          19997         41         127  2017-11-09 00:00:00         True   
19998          19998         87        2284  2017-04-14 00:00:00         True   
19999          19999          6        2764  2017-07-03 00:00:00        False   
20000          20000         11        1144  2017-09-22 00:00:00         True   

0     order_status           brand product_line product_class product_size  \
1         Approved           Solex     Standard        medium       medium   
2         Approved   Trek Bicycles     Standard        medium        large   
3         Approved      OHM Cycles     Standard           low       medium   
4         Approved  Norco Bicycles     Standard        medium       medium   
5         Approved  Giant Bicycles     Standard        medium        large   
...            ...             ...          ...           ...          ...   
19996     Approved      OHM Cycles     Standard          high       medium   
19997     Approved           Solex         Road        medium       medium   
19998     Approved      OHM Cycles     Standard        medium       medium   
19999     Approved      OHM Cycles     Standard          high       medium   
20000     Approved   Trek Bicycles     Standard        medium        small   

0     list_price standard_cost product_first_sold_date  
1          71.49         53.62                   41245  
2        2091.47        388.92                   41701  
3        1793.43        248.82                   36361  
4        1198.46         381.1                   36145  
5         1765.3        709.48                   42226  
...          ...           ...                     ...  
19996    2005.66        1203.4                   37823  
19997     416.98        312.74                   35560  
19998     1636.9         44.71                   40410  
19999     227.88        136.73                   38216  
20000    1775.81       1580.47                   36334  

[20000 rows x 13 columns]

In [122]:
transaction_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 1 to 20000
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   transaction_id           20000 non-null  object
 1   product_id               20000 non-null  object
 2   customer_id              20000 non-null  object
 3   transaction_date         20000 non-null  object
 4   online_order             19640 non-null  object
 5   order_status             20000 non-null  object
 6   brand                    19803 non-null  object
 7   product_line             19803 non-null  object
 8   product_class            19803 non-null  object
 9   product_size             19803 non-null  object
 10  list_price               20000 non-null  object
 11  standard_cost            19803 non-null  object
 12  product_first_sold_date  19803 non-null  object
dtypes: object(13)
memory usage: 2.0+ MB


In [123]:
transaction_df.describe()

0       transaction_id  product_id  customer_id     transaction_date  \
count            20000       20000        20000                20000   
unique           20000         101         3494                  364   
top                  1           0         2183  2017-02-14 00:00:00   
freq                 1        1378           14                   82   

0      online_order order_status  brand product_line product_class  \
count         19640        20000  19803        19803         19803   
unique            2            2      6            4             3   
top            True     Approved  Solex     Standard        medium   
freq           9829        19821   4253        14176         13826   

0      product_size  list_price  standard_cost  product_first_sold_date  
count         19803    20000.00       19803.00                    19803  
unique            3      296.00         103.00                      100  
top          medium     2091.47         388.92                    33879  
freq          12990      465.00         465.00                      234

In [124]:
#*Test: dropping first row:

#transaction_df.drop(index=transaction_df.index[0], axis=1, inplace=True)
#transaction_df

In [125]:
#*Test: dropping transaction_id and product_id Columns:

#transaction_df.drop(['transaction_id', 'product_id'], axis=1, inplace=True)
#transaction_df

## 2. Demographic DF:

In [126]:
demographic_df = pd.read_excel('KPMG_VI_New_raw_data_update_final.xlsx','CustomerDemographic')

/var/folders/hk/t9ycjb713h1fd831zgqd1zz40000gn/T/ipykernel_5833/1141691644.py:1: FutureWarning: Inferring datetime64[ns] from data containing strings is deprecated and will be removed in a future version. To retain the old behavior explicitly pass Series(data, dtype=datetime64[ns])
  demographic_df = pd.read_excel('KPMG_VI_New_raw_data_update_final.xlsx','CustomerDemographic')


In [131]:
demographic_df

customer_id               name  gender  \
88            720       Darrel Canet    Male   
89           1092     Katlin Creddon  Female   
90           3410    Merrili Brittin  Female   
91            658       Donn Bonnell    Male   
92           2413       Abbey Murrow    Male   
...           ...                ...     ...   
3995         2858    Benedicto Radki    Male   
3996         3173            Angelo     Male   
3997         3435    Stevena Allcock  Female   
3998         3949  Costa Sleightholm    Male   
3999         3997      Blanch Nisuis  Female   

      past_3_years_bike_related_purchases        DOB   age  \
88                                     67 1931-10-23  88.0   
89                                     56 1935-08-22  84.0   
90                                     93 1940-09-22  79.0   
91                                     38 1944-01-24  76.0   
92                                     27 1943-08-11  76.0   
...                                   ...        ...   ...   
3995                                    4 2002-01-09  18.0   
3996                                   55 2001-10-10  18.0   
3997                                   80 2002-01-15  18.0   
3998                                   24 2001-12-19  18.0   
3999                                   87 2001-07-13  18.0   

                       job_title job_industry_category     wealth_segment  \
88            Recruiting Manager                Retail  Affluent Customer   
89            VP Quality Control                Retail      Mass Customer   
90                           NaN              Property      Mass Customer   
91                Tax Accountant         Manufacturing  Affluent Customer   
92      Environmental Specialist         Manufacturing     High Net Worth   
...                          ...                   ...                ...   
3995          Recruiting Manager                   NaN      Mass Customer   
3996  Computer Systems Analyst I                Health      Mass Customer   
3997               Senior Editor                   NaN  Affluent Customer   
3998             Web Designer IV         Manufacturing     High Net Worth   
3999             Statistician II         Manufacturing     High Net Worth   

     deceased_indicator owns_car  tenure  
88                    N       No     6.0  
89                    N       No     5.0  
90                    N       No    16.0  
91                    N      Yes     8.0  
92                    N      Yes    17.0  
...                 ...      ...     ...  
3995                  N      Yes     1.0  
3996                  N      Yes     1.0  
3997                  N       No     1.0  
3998                  N      Yes     1.0  
3999                  N      Yes     1.0  

[3912 rows x 12 columns]

In [132]:
demographic_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3912 entries, 88 to 3999
Data columns (total 12 columns):
 #   Column                               Non-Null Count  Dtype         
---  ------                               --------------  -----         
 0   customer_id                          3912 non-null   int64         
 1   name                                 3912 non-null   object        
 2   gender                               3912 non-null   object        
 3   past_3_years_bike_related_purchases  3912 non-null   int64         
 4   DOB                                  3912 non-null   datetime64[ns]
 5   age                                  3912 non-null   float64       
 6   job_title                            3415 non-null   object        
 7   job_industry_category                3256 non-null   object        
 8   wealth_segment                       3912 non-null   object        
 9   deceased_indicator                   3912 non-null   object        
 10  owns_car   

In [133]:
demographic_df.describe()

customer_id  past_3_years_bike_related_purchases          age  \
count  3912.000000                          3912.000000  3912.000000   
mean   1999.828221                            49.039366    42.365542   
std    1154.903377                            28.768226    12.624297   
min       1.000000                             0.000000    18.000000   
25%    1002.750000                            24.000000    33.000000   
50%    2000.000000                            48.000000    42.000000   
75%    2997.250000                            73.000000    52.000000   
max    4000.000000                            99.000000    88.000000   

            tenure  
count  3912.000000  
mean     10.654652  
std       5.658898  
min       1.000000  
25%       6.000000  
50%      11.000000  
75%      15.000000  
max      22.000000

In [134]:
#filter 'age' so only data with age <90 is selected

demographic_df = demographic_df[demographic_df['age']<90]

In [135]:
demographic_df

customer_id               name  gender  \
88            720       Darrel Canet    Male   
89           1092     Katlin Creddon  Female   
90           3410    Merrili Brittin  Female   
91            658       Donn Bonnell    Male   
92           2413       Abbey Murrow    Male   
...           ...                ...     ...   
3995         2858    Benedicto Radki    Male   
3996         3173            Angelo     Male   
3997         3435    Stevena Allcock  Female   
3998         3949  Costa Sleightholm    Male   
3999         3997      Blanch Nisuis  Female   

      past_3_years_bike_related_purchases        DOB   age  \
88                                     67 1931-10-23  88.0   
89                                     56 1935-08-22  84.0   
90                                     93 1940-09-22  79.0   
91                                     38 1944-01-24  76.0   
92                                     27 1943-08-11  76.0   
...                                   ...        ...   ...   
3995                                    4 2002-01-09  18.0   
3996                                   55 2001-10-10  18.0   
3997                                   80 2002-01-15  18.0   
3998                                   24 2001-12-19  18.0   
3999                                   87 2001-07-13  18.0   

                       job_title job_industry_category     wealth_segment  \
88            Recruiting Manager                Retail  Affluent Customer   
89            VP Quality Control                Retail      Mass Customer   
90                           NaN              Property      Mass Customer   
91                Tax Accountant         Manufacturing  Affluent Customer   
92      Environmental Specialist         Manufacturing     High Net Worth   
...                          ...                   ...                ...   
3995          Recruiting Manager                   NaN      Mass Customer   
3996  Computer Systems Analyst I                Health      Mass Customer   
3997               Senior Editor                   NaN  Affluent Customer   
3998             Web Designer IV         Manufacturing     High Net Worth   
3999             Statistician II         Manufacturing     High Net Worth   

     deceased_indicator owns_car  tenure  
88                    N       No     6.0  
89                    N       No     5.0  
90                    N       No    16.0  
91                    N      Yes     8.0  
92                    N      Yes    17.0  
...                 ...      ...     ...  
3995                  N      Yes     1.0  
3996                  N      Yes     1.0  
3997                  N       No     1.0  
3998                  N      Yes     1.0  
3999                  N      Yes     1.0  

[3912 rows x 12 columns]

In [144]:
# Sort by 'past_3_years_bike_related_purchases'

demographic_df.sort_values(['past_3_years_bike_related_purchases'],ascending=False)

customer_id              name  gender  \
2044          561     Karin Burkill  Female   
3627         2013     Marge Girault  Female   
426          2354     Tammie Fisbey  Female   
3370          182     Lydon Dahlman    Male   
1942         2332      Jeff Goodere    Male   
...           ...               ...     ...   
2081         1195     Ally Folliott  Female   
1762         1042   Isacco Woodwind    Male   
3184         3843  Mahmoud Ligerton    Male   
2557         1291      Warner Zuker    Male   
437          2680       Noella Firk  Female   

      past_3_years_bike_related_purchases        DOB   age  \
2044                                   99 1977-08-15  42.0   
3627                                   99 1995-08-10  24.0   
426                                    99 1958-12-19  61.0   
3370                                   99 1992-11-20  27.0   
1942                                   99 1977-04-20  43.0   
...                                   ...        ...   ...   
2081                                    0 1978-03-22  42.0   
1762                                    0 1976-01-14  44.0   
3184                                    0 1989-06-06  31.0   
2557                                    0 1980-08-16  39.0   
437                                     0 1959-05-11  61.0   

                     job_title job_industry_category     wealth_segment  \
2044            Tax Accountant                   NaN      Mass Customer   
3627          Registered Nurse                Health  Affluent Customer   
426           Product Engineer                Retail      Mass Customer   
3370           Sales Associate    Financial Services  Affluent Customer   
1942             Developer III              Property     High Net Worth   
...                        ...                   ...                ...   
2081              VP Marketing    Telecommunications      Mass Customer   
1762    Nuclear Power Engineer         Manufacturing     High Net Worth   
3184  Senior Financial Analyst    Financial Services      Mass Customer   
2557          Quality Engineer                Health      Mass Customer   
437          Financial Analyst    Financial Services      Mass Customer   

     deceased_indicator owns_car  tenure  
2044                  N       No    12.0  
3627                  N      Yes    19.0  
426                   N       No    17.0  
3370                  N      Yes     7.0  
1942                  N      Yes    10.0  
...                 ...      ...     ...  
2081                  N       No    18.0  
1762                  N      Yes    11.0  
3184                  N      Yes    16.0  
2557                  N       No    12.0  
437                   N      Yes    14.0  

[3912 rows x 12 columns]

# Further Data Cleaning - Checking and working with missing value

## 1-A. Numerical Variables - Transaction DF Cleaning:

In [136]:
# Inspect missing values in the dataset
print(transaction_df.isnull().values.sum())

1542


In [137]:
# Replace the ' 's with NaN
transaction_df = transaction_df.replace(" ",np.NaN)

In [138]:
# Impute the missing values with mean imputation
transaction_df = transaction_df.fillna(transaction_df.mean())

/var/folders/hk/t9ycjb713h1fd831zgqd1zz40000gn/T/ipykernel_5833/4034010632.py:2: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  transaction_df = transaction_df.fillna(transaction_df.mean())


In [139]:
# Count the number of NaNs in the dataset (verify again after filling in the missing values)

print(transaction_df.isnull().values.sum())

788


In [140]:
# 请问，从 1542 missing values 到还有 788 missing values的原因？ 为什么不是到 0 missing values? 是因为还有文字版本的 missing values 吗？

## 1-B. Numerical Variables - Demographic DF Cleaning:

In [145]:
# Inspect missing values in the dataset
print(demographic_df.isnull().values.sum())

1153


In [146]:
# Replace the ' 's with NaN
demographic_df = transaction_df.replace(" ",np.NaN)

In [147]:
# Impute the missing values with mean imputation
demographic_df = transaction_df.fillna(transaction_df.mean())

/var/folders/hk/t9ycjb713h1fd831zgqd1zz40000gn/T/ipykernel_5833/1179130404.py:2: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  demographic_df = transaction_df.fillna(transaction_df.mean())


In [148]:
# Count the number of NaNs in the dataset (verify again after filling in the missing values)

print(demographic_df.isnull().values.sum())

788


## 2-A. Categorical Variables - Transaction DF Cleaning:

In [151]:
print(transaction_df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 1 to 20000
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   transaction_id           20000 non-null  int64  
 1   product_id               20000 non-null  int64  
 2   customer_id              20000 non-null  int64  
 3   transaction_date         20000 non-null  object 
 4   online_order             20000 non-null  object 
 5   order_status             20000 non-null  object 
 6   brand                    19803 non-null  object 
 7   product_line             19803 non-null  object 
 8   product_class            19803 non-null  object 
 9   product_size             19803 non-null  object 
 10  list_price               20000 non-null  float64
 11  standard_cost            20000 non-null  float64
 12  product_first_sold_date  20000 non-null  float64
dtypes: float64(3), int64(3), object(7)
memory usage: 2.0+ MB
None


In [ ]:
# 以上check info 显示的： 是不是这次再检查之后，剩下的还有 Null 的就都是 Categorical 的 missing values 了？

In [153]:
print(transaction_df.info())
for col in transaction_df.columns:
    
    # Check if the column is of object type
    if transaction_df[col].dtypes == 'object':
       
        # If the column is "object" type, then input with the most frequent value (find the most frequent value by-
        #- finding the value_counts, and the top row, which is 'index[0], is the most frequent value we want to find. 
        transaction_df[col] = transaction_df[col].fillna(transaction_df[col].value_counts().index[0])

# Count the number of NaNs in the dataset and print the counts to verify if it is '0' now. (which means 0 null value.)
print(transaction_df.isnull().values.sum())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 1 to 20000
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   transaction_id           20000 non-null  int64  
 1   product_id               20000 non-null  int64  
 2   customer_id              20000 non-null  int64  
 3   transaction_date         20000 non-null  object 
 4   online_order             20000 non-null  object 
 5   order_status             20000 non-null  object 
 6   brand                    19803 non-null  object 
 7   product_line             19803 non-null  object 
 8   product_class            19803 non-null  object 
 9   product_size             19803 non-null  object 
 10  list_price               20000 non-null  float64
 11  standard_cost            20000 non-null  float64
 12  product_first_sold_date  20000 non-null  float64
dtypes: float64(3), int64(3), object(7)
memory usage: 2.0+ MB
None
0


## 2-B. Categorical Variables - Demographic DF Cleaning:

In [152]:
print(demographic_df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 1 to 20000
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   transaction_id           20000 non-null  int64  
 1   product_id               20000 non-null  int64  
 2   customer_id              20000 non-null  int64  
 3   transaction_date         20000 non-null  object 
 4   online_order             20000 non-null  object 
 5   order_status             20000 non-null  object 
 6   brand                    19803 non-null  object 
 7   product_line             19803 non-null  object 
 8   product_class            19803 non-null  object 
 9   product_size             19803 non-null  object 
 10  list_price               20000 non-null  float64
 11  standard_cost            20000 non-null  float64
 12  product_first_sold_date  20000 non-null  float64
dtypes: float64(3), int64(3), object(7)
memory usage: 2.0+ MB
None


In [155]:
print(demographic_df.info())
for col in demographic_df.columns:
    
    # Check if the column is of object type
    if demographic_df[col].dtypes == 'object':
       
        # If the column is "object" type, then input with the most frequent value (find the most frequent value by-
        #- finding the value_counts, and the top row, which is 'index[0], is the most frequent value we want to find. 
        demographic_df[col] = demographic_df[col].fillna(demographic_df[col].value_counts().index[0])

# Count the number of NaNs in the dataset and print the counts to verify if it is '0' now. (which means 0 null value.)
print(demographic_df.isnull().values.sum())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 1 to 20000
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   transaction_id           20000 non-null  int64  
 1   product_id               20000 non-null  int64  
 2   customer_id              20000 non-null  int64  
 3   transaction_date         20000 non-null  object 
 4   online_order             20000 non-null  object 
 5   order_status             20000 non-null  object 
 6   brand                    19803 non-null  object 
 7   product_line             19803 non-null  object 
 8   product_class            19803 non-null  object 
 9   product_size             19803 non-null  object 
 10  list_price               20000 non-null  float64
 11  standard_cost            20000 non-null  float64
 12  product_first_sold_date  20000 non-null  float64
dtypes: float64(3), int64(3), object(7)
memory usage: 2.0+ MB
None
0


# Originally, we can see that we had 1542 and 1153 null values from the two data frames.  We treated them with inputting mean values, as well as the most frequent categorical values per datatype.  Now, as we have completed our data cleaning and understanding, we will commence the Cohort Analysis.

# EDA:

# Assigned the cohorts and calculated the monthly offset

In [158]:
# For example, this is to find the 1st row of the column 'transaction_date'
transaction_df['transaction_date'][1]


Timestamp('2017-02-25 00:00:00')

In [159]:
# For exapmle, this is to extract the YEAR and MONTH from the first row of the column 'transaction_date'.
# Since we just want to find the Year and Month of each transaction to set up a column representing the Cohort Month (Starting Month)-
#-we manually set date as "1". (the last attribute in the code)

dt.datetime(transaction_df['transaction_date'][1].year, transaction_df['transaction_date'][1].month, 1)

datetime.datetime(2017, 2, 1, 0, 0)